File: ruby/pile.rb
class Pile
  attr_reader :sample, :remaining

  def initialize(card_factory, card_count)
    @factory = card_factory
    @remaining = card_count
    @sample = card_factory.new
  end

  def empty?
    @remaining <= 0
  end

  def push
    @remaining += 1
  end

  def pop
    fail "#{@sample} pile is empty" if empty?
    @remaining -= 1
    @factory.new
  end

  def to_s
    "#{@sample} [#{@remaining}]"
  end
end

File: python/pile.py
class Pile:
    def __init__(self, factory, count):
        self.factory = factory
        self.remaining = count
        self.sample = self.factory()

    def empty(self):
        return self.remaining <= 0

    def push(self):
        self.remaining += 1

    def pop(self):
        assert not self.empty, '{} pile is empty'.format(self.sample.name)
        self.remaining -= 1
        return self.factory()
